source/extensions/filters/http/lua/lua_filter.h

```diff
@@ -376,6 +380,21 @@ class Filter : public Http::StreamFilter, Logger::Loggable<Logger::Id::lua> {
   StreamHandleRef request_stream_wrapper_;
   StreamHandleRef response_stream_wrapper_;
   bool destroyed_{};
+
+  // These coroutines used to be owned by the stream handles. After investigating #3570, it
+  // became clear that there is a circular memory reference when a coroutine yields. Basically,
+  // the coroutine holds a reference to the stream wrapper. I'm not completely sure why this is,
+  // but I think it is because the yield happens via a stream handle method, so the runtime must
+  // hold a reference so that it can return out of the yield through the object. So now we hold
+  // the coroutine references at the same level as the stream handles so that when the filter is
+  // destroyed the circular reference is broken and both objects are cleaned up.
+  //
+  // Note that the above explanation probably means that we don't need to hold a reference to the
+  // coroutine at all and it would be taken care of automatically via a runtime internal reference
+  // when a yield happens. However, given that I don't fully understand the runtime internals, this
+  // seems like a safer fix for now.
```

Comment:
My $0.02 is that this is a bit unsatisfactory in terms of explanation; we're basically introducing a change in object ownership without a clear narrative as to why it's necessary. I think what you describe is logically plausible, but I'd ask if we could figure out what the underlying truthiness is here.